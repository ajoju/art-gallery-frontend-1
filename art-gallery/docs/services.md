---
title: Services
---

# Services

## AuthService

Used to create the auth header for API requests which require authentication.

## UserService

* Handles user login and registration requests by sending API requests to the relevant end points.
* Receives token data and stores relevant information in the account store.

### Methods

#### `Login()`

Authenticates user information gathered using the LoginComponent's form data, and creates user data if login is successful, including:
* The token, for making authenticated requests
* The user's given name for displaying a message in the navigation component
* The expiry of the token to enable logout when tokens expire
* The user's role

#### `Logout()`

Logs a user out by deleting the stored user information from local storage and reloading the page.

#### `handleLoginResponse()`

Receives the request from the Login() method, sends it to the API and handles the data received in the API response.

#### `SignUp()`

Submits user information gathered using the registration form data.

#### `handleSignupResponse()`

Receives the request from the SignUp() method, sends it to the API and handles the data received in the API response.

<br>
<br>

