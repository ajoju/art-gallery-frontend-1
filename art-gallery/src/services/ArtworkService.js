// Fetch artworks from database.
export async function getArtworks() {
  const response = await fetch('/api/artworks')
  return await response.json();
}

// Fetch artwork of the day.
export async function getArtworkOfTheDay() {
  const response = await fetch('/api/artworks/of-the-day')
  return await response.json();
}