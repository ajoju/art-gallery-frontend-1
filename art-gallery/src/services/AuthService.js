// Service to handle the auth header
export function authHeader() {
    let user = JSON.parse(localStorage.getItem('user'));

    if (user) {
      var token = user.token;
      token = token.replace(/^"(.*)"$/, '$1');
      return 'Bearer ' + token ;
    } else {
      return {};
    }
  }
