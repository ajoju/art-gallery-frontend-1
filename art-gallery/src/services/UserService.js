import VueJwtDecode from 'vue-jwt-decode';

// Service to handle logging in and out
export const userService = {
    Login,
    Logout,
    SignUp
};

async function Login(email, password) {
    const request = {
      method: "POST",
      headers: { "Content-Type": "application/json"},
      body: JSON.stringify({ email: email, password: password})
    };

    await fetch('/api/users/login/', request)
        .then(handleLoginResponse)
        .then(data => {
            const decoded = VueJwtDecode.decode(data);
            localStorage.setItem('user',
            JSON.stringify({
                token: data,
                name: decoded.given_name,
                expiry: decoded.exp,
                role: decoded.role
            })
            );
            return data
        })
}

function Logout() {
    // remove user from local storage to log user out
    localStorage.removeItem('user');
    location.reload(true);
}

function handleLoginResponse(response) {
    return response.text().then(text => {
        const data = text && JSON.parse(text);
        if (!response.ok) {
            if (response.status === 401) {
                // auto logout if 401 response returned from api
                Logout();
            }
            const error = (data && data.message) || response.statusText;
            return Promise.reject(error);
        }
        return data;
    });
}

// Register a new account using the sign up api
export async function SignUp(firstName, lastName, email, password) {
    const request = {
      method: "POST",
      headers: { "Content-Type": "application/json"},
      body: JSON.stringify({ firstName: firstName, lastName:lastName, email: email, password: password})
    }
    await fetch('/api/users/signup/', request)
        .then(handleSignupResponse);
}

function handleSignupResponse(response) {
    return response.text().then(text => {
        const data = text && JSON.parse(text);
        if (!response.ok) {
            if (response.status === 401) {
                // auto logout if 401 response returned from api
                Logout();
            }

            const error = (data && data.message) || response.statusText;
            return Promise.reject(error);
        }
        return data;
    });
}
