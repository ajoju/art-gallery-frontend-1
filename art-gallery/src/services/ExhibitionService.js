import { authHeader } from '@/services/AuthService'

// Fetch exhibitions from database.
export async function getExhibitions() {
  const response = await fetch('/api/exhibitions')
  return await response.json();
}

// Post exhibition to database.
export async function postExhibition(name, description, backgroundImageUrl, startDate, endDate) {
  const request = {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: authHeader()
    },
    body: JSON.stringify({ name: name, description: description, backgroundImageUrl: backgroundImageUrl, startDate: startDate, endDate: endDate})
  }
    let response = await fetch('/api/exhibitions', request)
    let responseJSON = response.json()

    return responseJSON
}
