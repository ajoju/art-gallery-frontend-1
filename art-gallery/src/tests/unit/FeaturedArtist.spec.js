import { mount } from "@vue/test-utils";
import FeaturedArtist from "@/components/FeaturedArtist.vue";
import CardComponent from "@/components/CardComponent.vue";
import { getArtistOfTheDay } from "@/services/ArtistService";
import { vi } from "vitest";

vi.mock("@/services/ArtistService", () => ({
  getArtistOfTheDay: vi.fn(),
}));

describe("FeaturedArtist", () => {
  let wrapper;

  const mockArtist = {
    displayName: "Artist Name",
    profileImageUrl: "https://example.com/image.jpg",
    yearOfBirth: 1980,
    placeOfBirth: "New York",
  };

  beforeEach(async () => {
    getArtistOfTheDay.mockResolvedValue(mockArtist);

    wrapper = mount(FeaturedArtist);

    // Wait for fetchArtistOfTheDay() to complete.
    await wrapper.vm.$nextTick();
  });

  it("renders correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders the CardComponent", () => {
    const cardComponent = wrapper.findComponent(CardComponent);
    expect(cardComponent.exists()).toBe(true);
  });

  it("passes the correct props to CardComponent", () => {
    const cardComponent = wrapper.findComponent(CardComponent);

    expect(cardComponent.props().subheading).toBe(mockArtist.displayName);
    expect(cardComponent.props().imageURL).toBe(mockArtist.profileImageUrl);
    expect(cardComponent.props().detail1).toBe(
      `Born: ${mockArtist.yearOfBirth}, ${mockArtist.placeOfBirth}`
    );
  });
});
