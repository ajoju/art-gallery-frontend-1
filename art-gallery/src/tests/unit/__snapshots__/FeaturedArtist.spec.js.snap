// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FeaturedArtist > renders correctly 1`] = `
"<div data-v-f967684d=\\"\\" class=\\"card-div\\">
  <h1 data-v-f967684d=\\"\\" class=\\"card-heading\\"></h1><img data-v-f967684d=\\"\\" class=\\"card-img\\" src=\\"https://example.com/image.jpg\\">
  <h2 data-v-f967684d=\\"\\" class=\\"card-subheading\\">Artist Name</h2>
  <h4 data-v-f967684d=\\"\\" class=\\"detail\\">Born: 1980, New York</h4>
  <h4 data-v-f967684d=\\"\\" class=\\"detail\\"></h4>
  <h4 data-v-f967684d=\\"\\" class=\\"detail\\"></h4>
  <h4 data-v-f967684d=\\"\\" class=\\"detail-description\\"></h4>
</div>"
`;
